platform/x11/tts_linux.cpp

```diff
@@ -166,20 +171,15 @@ Array TTS_Linux::get_voices() const {
 	_THREAD_SAFE_METHOD_
 
 	ERR_FAIL_COND_V(!synth, Array());
+	const_cast<TTS_Linux *>(this)->_load_voices();
+
 	Array list;
-	SPDVoice **voices = spd_list_synthesis_voices(synth);
-	if (voices != nullptr) {
-		SPDVoice **voices_ptr = voices;
-		while (*voices_ptr != nullptr) {
-			Dictionary voice_d;
-			voice_d["name"] = String::utf8((*voices_ptr)->name);
-			voice_d["id"] = String::utf8((*voices_ptr)->name);
-			voice_d["language"] = String::utf8((*voices_ptr)->language) + "_" + String::utf8((*voices_ptr)->variant);
-			list.push_back(voice_d);
-
-			voices_ptr++;
-		}
-		free_spd_voices(voices);
+	for (OrderedHashMap<String, VoiceInfo>::ConstElement E = voices.front(); E; E = E.next()) {
+		Dictionary voice_d;
+		voice_d["name"] = E.key();
+		voice_d["id"] = E.key();
+		voice_d["language"] = E.get().language + "_" + E.get().variant;
+		list.push_back(voice_d);
```

Comment:
If this hashmap is never used, can't you just cache the Array directly instead of the hashmap?

Reply:
If someone chooses to backport the language-specific word splitting, then they will want this.   This also matches up better with the 4.x code, making it easier to backport future bug fixes.  

If you want, I can change it, but I would generally prefer to keep the 3.x code similar to the 4.x code, rather than focusing on a small efficiency.

Reply:
imo it depends how often this function `get_voices()` is called. If it is a one off, then there's no great problem - but in that case, there would be no need to cache you would have thought.

The main concern here is that it is building a new Array of Dictionaries each time `get_voices()` is called, where it could just cache the Array.

But looking at the linked issue, #77754 , the problem in 4.x seems to be that loading happens each time a speech event occurs, which is not the case in 3.x. So if `get_voices()` was only a one off (I have no idea if this is the case), there wouldn't seem much benefit to caching. :thinking: